keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

```diff
@@ -0,0 +1,388 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The main goal of this enhancement is to authoritatively identify the Operating System
+on which a pod would like to run.
+
+
+## Motivation
+
+Identifying the OS of the pods during the API Server admission time is crucial to apply
+appropriate security constraints to the pod. Without proper identification, some admission
+plugins may apply unnecessary security constraints to the pods or in the
+worst case, don't apply security constraints at all.
+
+
+### Goals
+
+Admission plugins and/or Admission Webhooks identifying the Operating System on which
+pod intends to run.
+
+### Non-Goals
+
+
+- Validating if the pods have appropriate security constraints.
+- Interaction with container runtimes.
+
+## Proposal
+
+
+Identifying the Windows pods during the scheduling phase can be done
+in the following ways:
+- Based on [nodeSelector and tolerations](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec with 
+  Windows node specific labels.
+- Based on [runtimeclasses](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec
+
+The runtimeclass is a higher level abstraction which gets translated again to nodeSelectors+tolerations. While the 
+nodeSelector with reserved OS label is good enough, it can be viewed as an side-effect of scheduling constraints 
+and need not necessarily express the user intention fully. So, we propose to add a new field to the pod spec called 
+os to identify the OS on which pod intends to run.
+
+```go
+type PodSpec struct {
+  // If specified, indicates the type of OS on which pod intends
+  // to run.  This field is alpha-level and is only honored by servers that 
+  // enable the IdentifyOS feature. This is used to help identifying the
+  // OS authoritatively during the API server admission.
+  // +optional
+  OS *OS
+}
+
+// OS has information on the type of OS. 
+// We're making this a struct for possible future expansion.
+type OS struct {
+  // Name of the OS. The values supported are available at:
+  // https://github.com/opencontainers/runtime-spec/blob/master/config.md#platform-specific-configuration
```

Comment:
We support all of these?  solaris?

When the API is created, I think a strict enumeration is required.

Reply:
current node platforms:
```
# The node platforms we build for
readonly KUBE_SUPPORTED_NODE_PLATFORMS=(
  linux/amd64
  linux/arm
  linux/arm64
  linux/s390x
  linux/ppc64le
  windows/amd64
)
```

Reply:
that said, I'm not necessarily opposed to accepting more OS values, I just think that means more work to define API validation and PodSecurity treatment of each enumerated value

when we add validation for this, we should probably bake in toleration of an existing unknown value on an update, and explicitly note that future values could be added in the future, and how a client should handle an unknown enum value (xref https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field)

Reply:
Ahh I see, I was doing something like this in the validation stage - https://github.com/kubernetes/kubernetes/pull/104693/commits/852f44253137ddfba8209eee1385827f9657f835#diff-c713e8919642d873fdf48fe8fb6d43e5cb2f53fd601066ff53580ea655948f0dR6158. I believe I need to add some toleration for the unknown updates.